File: Leetcode100/leetcode100_python/MaximumSubarray2.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # lsum rsum isum msum
        def helper(left, right):
            if left >= right:
                return (nums[left], nums[left], nums[left], nums[left])
            mid = (left + right) // 2
            lsub = helper(left, mid)
            rsub = helper(mid + 1, right)

            lsum = max(lsub[2] + rsub[0], lsub[0])
            rsum = max(rsub[2] + lsub[1], rsub[1])
            isum = lsub[2] + rsub[2]
            msum = max(lsub[3], rsub[3], lsub[1] + rsub[0])
            return (lsum, rsum, isum, msum)
        return helper(0, len(nums) - 1)[3]

if __name__ == "__main__":
    nums = [int(num) for num in input().split()]
    solution = Solution()
    print(solution.maxSubArray(nums))

File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/maximum_subarray/MaximumSubarray2.java
package main.java.com.jiang.leetcode100.maximum_subarray;

import java.util.Arrays;
import java.util.Scanner;

class Solution2 {
    public int maxSubArray(int[] nums) {
        return helper(nums, 0, nums.length - 1);
    }

    public int helper(int[] nums, int left, int right) {
        if(left >= right) return nums[left];
        int mid = (left + right) >> 1;
        int lmax = helper(nums, left, mid - 1);
        int rmax = helper(nums, mid + 1, right);
        int sum = nums[mid], t = nums[mid];
        for(int i = mid + 1; i <= right; i++) {
            t += nums[i];
            sum = Math.max(sum, t);
        }
        t = sum;
        for(int i = mid - 1; i >= left; i--) {
            t += nums[i];
            sum = Math.max(sum, t);
        }
        return Math.max(Math.max(lmax, rmax), sum);
    }
}

public class MaximumSubarray2 {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String str = scanner.nextLine();
        scanner.close();
        String[] nums_str = str.split(" ");
        int[] nums = new int[nums_str.length];
        for(int i = 0; i < nums_str.length; i++) {
            nums[i] = Integer.parseInt(nums_str[i]);
        }

        System.out.println(new Solution2().maxSubArray(nums));
    }
}